shared/modules/caip-stream.ts

```diff
@@ -0,0 +1,71 @@
+import { isObject } from '@metamask/utils';
+import { pipeline, Duplex } from 'readable-stream';
+
+class Substream extends Duplex {
+  parent: Duplex;
+
+  constructor(parent: Duplex) {
+    super({ objectMode: true });
+    this.parent = parent;
+  }
+
+  _read() {
+    return undefined;
+  }
+
+  _write(
+    value: unknown,
+    _encoding: BufferEncoding,
+    callback: (error?: Error | null) => void,
+  ) {
+    this.parent.push({
+      type: 'caip-x',
```

Comment:
Are our ObjectMultiplex streams setup to expect `name` instead of `type`? You mentioned this in a message and I see [ObjectMultiplex uses `name`](https://github.dev/MetaMask/object-multiplex)

Reply:
that is correct. multiplexed streams use `name` to differentiate streams. I'm not convinced that `name` is a good name for the equivalent property in the CAIP envelope

Reply:
But don't we have to inherit it nonetheless?

Reply:
Ok @jiexi has explained to me why this isn't a problem. When I asked this I had forgotten that we're creating a whole separate stream not a substream to the existing multiplexed stream which allows us to rename this property.